scheduler/src/cook/compute_cluster.clj

```diff
@@ -156,3 +159,312 @@
+
+;TODO see if this is ok or need a better way
+(def scheduler-promise (promise))
+(def exit-code-syncer-state-promise (promise))
+
+(defn initialize-cluster!
+  "Create and initialize a ComputeCluster"
+  [config]
+  (let [cluster-definition-template ((config/compute-cluster-templates) (:template config))
+        _ (when-not cluster-definition-template (throw (ex-info "Attempting to create cluster with unknown template" {:config config})))
+        factory-fn (:factory-fn cluster-definition-template)
+        _ (when-not factory-fn (throw (ex-info "Template for cluster has no factory-fn" {:config config})))
+        resolved (cook.util/lazy-load-var factory-fn)
+        full-cluster-config (merge (:config cluster-definition-template) config)
+        cluster (resolved full-cluster-config {:exit-code-syncer-state @exit-code-syncer-state-promise})]
+    (initialize-cluster cluster (:pool-name->fenzo @scheduler-promise))))
+
+(defn execute-update!
+  "Attempt to execute a valid cluster configuration update.
+  Change cluster state if there is an existing cluster in memory.
+  Create clusters if there is an active update an no corresponding cluster in memory. This can happen at startup or
+  when a brand new cluster is added.
+  Reflect changes in the database."
+  [conn
+   {:keys [valid? changed? active?] {:keys [name state state-locked?] :as goal-config} :goal-config}
+   current-in-mem-configs]
+  {:pre [valid?]}
+  (try
+    (let [{:keys [state-atom state-locked?-atom] :as cluster} (@cluster-name->compute-cluster-atom name)
+          current-in-mem-config (current-in-mem-configs name)]
+      ; TODO: keep these checks? check some other way?
+      (when (and current-in-mem-config (not (and cluster state-atom state-locked?-atom)))
+        (throw (ex-info "We know an in-memory config but we don't know the corresponding in-memory cluster.
+        This should never happen, since cluster-name->compute-cluster-atom should be locked"
+                        {:current-in-mem-config current-in-mem-config
+                         :cluster cluster})))
+      (when (and (or cluster state-atom state-locked?-atom) (not current-in-mem-config))
+        (throw (ex-info "We know an in-memory cluster but we don't know the corresponding in-memory config.
+        This should never happen, since cluster-name->compute-cluster-atom should be locked"
+                        {:current-in-mem-config current-in-mem-config
+                         :cluster cluster})))
+      (if current-in-mem-config
+        (do
+          (reset! state-atom state)
+          (reset! state-locked?-atom state-locked?))
+        (when active? (initialize-cluster! goal-config))))
+    (when changed?
+      ; :db.unique/identity gives upsert behavior (update on insert), so we don't need to specify an entity ID when updating
+      @(d/transact conn [(assoc (compute-cluster-config->compute-cluster-config-ent goal-config)
```

Comment:
I mentioned this earlier, but we should do the db updates first, before we modify the in-memory state. That way the Db is the system of record and we avoid various races.

Reply:
I don't want to not be able to change cluster state if there's a database issue
I'm ok doing a try catch on each one separately

Reply:
things that are changing the clusters don't rely on the old state, they just say what they want the new state to be

I'm fine re-ordering if you think that the db failing is fatal anyway

Reply:
Cool. Thanks.